Move verify_delegate() to Root/Targets

jku opened the pull request:
Fixes #2361 

**Description of the changes being introduced by the pull request**:

* Add an internal `_Delegator` class that implements verify_delegate()
* Inherit that class in Root, Targets: the internal class is just an implementation detail for sharing the implementation, not something users of the library should care about
* Document Metadata.verify_delegate() as deprecated (and use the new implementation internally)
* Use the new method in tests, examples and ngclient

This will make both ngclient and external repository code slightly easier to read and write (but note that the ngclient cleanups are not included here as they touch quite a lot of lines even if they are simple).

There are several ways to add verify_delegate() to Root and Targets, I think this helper class is a decent compromise between modularity and simplicity: 
* the _Delegator class isn't intended to be part of a longer "inheritance path" (and does not have a matching element in the metadata format): it just provides an extra method for Root and Targets. If there's a naming style for this style of inheritance I'm happy to use it
* _Delegator is internal: users of the library don't need to care about it

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

jku committed:
Move verify_delegate() to Root/Targets

This makes logical sense and makes a lot of code using
verify_delegate() a little easier since there is no need to keep a
reference to the containing metadata anymore.

The implementation is in practice in a new class but that's an
implementation detail that allows sharing between Targets and Root.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
ngclient,tests,examples: Use new verify_delegate()

Avoid Metadata.verify_delegate() now that it's deprecated.

Note that this commit does not try to make any code cleanups
that are now possible: this is the minimal change to use the new
API.

Future improvements can make code in TrustedMetadataSet and
Updater slightly easier to read: as an example there's no need for
TrustedMetadataSet to actually store or expose actual Metadata in its
cache -- Signed is all that's needed.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku commented:
> If there's a naming style for this style of inheritance I'm happy to use it

A "Mixin" is likely what I'm trying to remember here?

jku committed:
metadata: Rename _Delegator to _DelegatorMixin

Make it clearer that this is not part of the main inheritance path.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku commented:
I think this is ready for review. I have a branch [cleanup-after-verify-delegate-move](https://github.com/jku/python-tuf/commits/cleanup-after-verify-delegate-move) with changes based on this PR... but they are not required so I'll do those separately.

jku commented on tuf/api/metadata.py:
```diff
@@ -1926,7 +1962,7 @@ def get_delegated_role(self, delegated_role: str) -> Role:
         """Return the key object for the given keyid.
 
-        Raises ValueError if keyid is not found.
+        Raises ValueError if key is not found.
```
this change just makes the two implementations have the same docstring.

lukpueh replied:
You could actually remove docstrings for `get_key` and `get_delegated_role` in `Targets` and `Root`, since they can be reused from their definitions in the parent class. Pylint would need to be told to not complain. So maybe not worth it, the docstrings are short enough.

jku replied:
I was sure that wouldn't work with sphinx but I tried and you are right.

jku commented on tuf/api/metadata.py:
```diff
@@ -675,7 +635,83 @@ def to_dict(self) -> Dict[str, Any]:
+        """
+        raise NotImplementedError
+
+    def verify_delegate(
+        self,
+        delegated_role: str,
+        delegated_metadata: "Metadata",
+        signed_serializer: Optional[SignedSerializer] = None,
+    ) -> None:
```
Maybe a meaningful note here is that this makes `Targets` and `Root` "aware" of `Metadata` now.

The alternative would be for the arguments to include payload bytes and signatures so something like 

    def verify_delegate(
        self,
        delegated_role: str,
        payload: bytes,
        signatures: Dict[str, Signature]
    )

Unfortunately I'm not sure how one would make that usable in calling code

jku committed:
Metadata API: Annotation syntax tweak

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -396,56 +396,16 @@ def verify_delegate(
-                required threshold of keys for ``role_name``.
-            ValueError: no delegation was found for ``delegated_role``.
-            TypeError: called this function on non-delegating metadata class.
+        .. deprecated:: 2.2.0
```
Is this the correct version number?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -675,7 +635,83 @@ def to_dict(self) -> Dict[str, Any]:
+        signed_serializer: Optional[SignedSerializer] = None,
+    ) -> None:
+        """Verify that ``delegated_metadata`` is signed with the required
+        threshold of keys for the delegated role ``delegated_role``.
```
Off-topic and not important, but here's a chance for leaner docs, which also makes it consistent with the first part of the sentence.
```suggestion
        threshold of keys for the ``delegated_role``.
```

lukpueh commented on tests/test_api.py:
```diff
@@ -345,44 +345,83 @@ def test_metadata_verify_delegate(self) -> None:
         with self.assertRaises(ValueError):
             role2.verify_delegate("role1", role1)
 
+    # pylint: disable=too-many-locals,too-many-statements
```
I think this disables pylint downwards on class scope and not on function scope. Probably not what you intended, but also no big deal.

lukpueh reviewed (changes requested):
This change makes sense. I pointed out some nits that you may ignore. The only thing that should be changed is `test_api.TestMetadata.test_metadata_verify_delegate`, which still uses `Metadata.verify_delegate`. Otherwise, the tests pass even without that method.

Also, I think the version number in the deprecation notice is wrong (see inline comment).

jku committed:
Metadata API: Improve docstrings

Remove duplicate docstrings: these are already documented in
_DelegatorMixin and sphinx will find them there.

Tweak a few other strings to remove duplication in the sentence.

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

Metadata API: Improve dosctrings

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
tests: Move lint disable to inside block as intended

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku committed:
Metadata API: Bump deprecation version to next likely candidate

Signed-off-by: Jussi Kukkonen <jkukkonen@google.com>

jku commented:
> `test_api.TestMetadata.test_metadata_verify_delegate`, which still uses `Metadata.verify_delegate`

I left a small test in place to make sure I didn't break that one -- just enough testing to make sure it's connected to the actual implementation in Signed. 

`test_signed_verify_delegate()` contains all of the tests that existed before.

jku commented:
```
lint: commands[1]> isort --check --diff tuf examples tests verify_release
ERROR: /home/runner/work/python-tuf/python-tuf/verify_release Imports are incorrectly sorted and/or formatted.
--- /home/runner/work/python-tuf/python-tuf/verify_release:before	2023-05-15 08:14:43.101068
+++ /home/runner/work/python-tuf/python-tuf/verify_release:after	2023-05-15 08:15:05.396047
@@ -19,9 +19,8 @@
 from typing import Optional
 
 try:
+    import build as _
     import requests
-
-    import build as _
 except ImportError:
     print("Error: verify_release requires modules 'requests' and 'build':")
     print("    pip install requests build")
lint: exit 1 (0.27 seconds) /home/runner/work/python-tuf/python-tuf> isort --check --diff tuf examples tests verify_release pid=1912
```

The code that is failing on CI is exactly the code that my isort is suggesting I use.

Same exact isort version
 
 (╯°□°)╯︵ ┻━┻

jku commented:
I've dropped the commit that would make lint work on my machine... CI doesn't seems to mind :shrug: 

Let's maybe not merge this yet though: I still feel like we might be designing ourselves into a corner wrt DSSE.

Marking draft while thinking about it...

Commit:
Metadata API: add Metadata.signed_bytes property

- Add shortcut to canonical json representation of self.signed
- Use in tests and Metadata.sign
- Do not use in _Delegator.verify_delegate (will be updated in subsequent
  commit).

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Metadata API: make new verify_delegate unaware of Metadata

Change new _Delegator.verify_delegate to take payload bytes and
signatures instead of a Metadata object and a payload serializer.

This allows using verify_delegate for payloads that do not come in
a Metadata container, but e.g. in a DSSE envelope (see #2385).

Usage becomes a bit more cumbersome, but still feels reasonable with the
recently added shortcut for default canonical bytes representation of
Metadata.signed.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
> The alternative would be for the arguments to include payload bytes and signatures so something like
> 
> ```
> def verify_delegate(
>     self,
>     delegated_role: str,
>     payload: bytes,
>     signatures: Dict[str, Signature]
> )
> ```
> 
> Unfortunately I'm not sure how one would make that usable in calling code

I have a patch for this on top of your (rebased) branch. It also adds an alias for default canonical signed bytes to Metadata. This makes the calling code only slightly more complicated: https://github.com/theupdateframework/python-tuf/compare/635a2870bde3c44b4eab9e66be61c77e08e47285...lukpueh:move-verify-delegate-v2

I think this is a reasonable path towards #2385. @jku, feel free to take over my branch for this PR, or let me know if you want me to force push.

lukpueh commented:
@jku, you may want to approve this pr too, now that it's a collaboration.

jku commented:
It looks like original author cannot click "approve"... but I think this looks correct and reasonable.
